apps/lib/opt.c

```diff
@@ -367,52 +367,73 @@ void print_format_error(int format, unsigned long flags)
     (void)opt_format_error(format2str(format), flags);
 }
 
+/*
+ * Bump count of provider options, return new value.
+ * This function should be cleaner, and be in app_provider.c, but
+ * the tests only link against this file so it has to be here and
+ * this API is clunky but workable.
+ */
+int opt_incr_provider_options(int x)
+{
+    static int count;
+
+    return count += x;
+}
+
 /* Parse a cipher name, put it in *EVP_CIPHER; return 0 on failure, else 1. */
-int opt_cipher_silent(const char *name, EVP_CIPHER **cipherp)
+int opt_cipher_silent(OSSL_LIB_CTX *libctx,
+                      const char *name, const char *propq,
+                      EVP_CIPHER **cipherp)
 {
     EVP_CIPHER_free(*cipherp);
 
     ERR_set_mark();
-    if ((*cipherp = EVP_CIPHER_fetch(NULL, name, NULL)) != NULL
-        || (*cipherp = (EVP_CIPHER *)EVP_get_cipherbyname(name)) != NULL) {
+    if ((*cipherp = EVP_CIPHER_fetch(libctx, name, propq)) != NULL
+            || (!opt_incr_provider_options(0)
+                && (*cipherp =
+                        (EVP_CIPHER *)EVP_get_cipherbyname(name)) != NULL)) {
```

Comment:
Looking back, I find this solution a bit weird.  What happens if someone uses *both* `-provider` and `-engine`?  That's a valid combination, they don't exclude each other!

Reply:
based on the documentation in this PR, it would seem that it would be an error to have both engine and providers.

Reply:
In that case, I would say that this PR isn't quite right.